source/ocproute.go

```diff
@@ -0,0 +1,282 @@
+
+	targets := getTargetsFromTargetAnnotation(ocpRoute.Annotations)
+
+	if len(targets) == 0 {
+		targets = targetsFromOcpRouteStatus(ocpRoute.Status)
+	}
+
+	providerSpecific, setIdentifier := getProviderSpecificAnnotations(ocpRoute.Annotations)
+
+	var endpoints []*endpoint.Endpoint
+	// splits the FQDN template and removes the trailing periods
+	hostnameList := strings.Split(strings.Replace(hostnames, " ", "", -1), ",")
+	for _, hostname := range hostnameList {
+		hostname = strings.TrimSuffix(hostname, ".")
+		endpoints = append(endpoints, endpointsForHostname(hostname, targets, ttl, providerSpecific, setIdentifier)...)
+	}
+	return endpoints, nil
+}
+
+func (ors *ocpRouteSource) filterByAnnotations(ocpRoutes []*routeapi.Route) ([]*routeapi.Route, error) {
+	labelSelector, err := metav1.ParseToLabelSelector(ors.annotationFilter)
+	if err != nil {
+		return nil, err
+	}
+	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	// empty filter returns original list
+	if selector.Empty() {
+		return ocpRoutes, nil
+	}
+
+	filteredList := []*routeapi.Route{}
+
+	for _, ocpRoute := range ocpRoutes {
+		// convert the Route's annotations to an equivalent label selector
+		annotations := labels.Set(ocpRoute.Annotations)
+
+		// include ocpRoute if its annotations match the selector
+		if selector.Matches(annotations) {
+			filteredList = append(filteredList, ocpRoute)
+		}
+	}
+
+	return filteredList, nil
+}
+
+func (ors *ocpRouteSource) setResourceLabel(ocpRoute *routeapi.Route, endpoints []*endpoint.Endpoint) {
+	for _, ep := range endpoints {
+		ep.Labels[endpoint.ResourceLabelKey] = fmt.Sprintf("route/%s/%s", ocpRoute.Namespace, ocpRoute.Name)
+	}
+}
+
+// endpointsFromOcpRoute extracts the endpoints from a OpenShift Route object
+func endpointsFromOcpRoute(ocpRoute *routeapi.Route, ignoreHostnameAnnotation bool) []*endpoint.Endpoint {
+	var endpoints []*endpoint.Endpoint
+
+	ttl, err := getTTLFromAnnotations(ocpRoute.Annotations)
+	if err != nil {
+		log.Warn(err)
+	}
+
+	targets := getTargetsFromTargetAnnotation(ocpRoute.Annotations)
+
+	if len(targets) == 0 {
+		targets = targetsFromOcpRouteStatus(ocpRoute.Status)
+	}
+
+	providerSpecific, setIdentifier := getProviderSpecificAnnotations(ocpRoute.Annotations)
+
+	if host := ocpRoute.Spec.Host; host != "" {
+		endpoints = append(endpoints, endpointsForHostname(host, targets, ttl, providerSpecific, setIdentifier)...)
+	}
+
+	// Skip endpoints if we do not want entries from annotations
+	if !ignoreHostnameAnnotation {
+		hostnameList := getHostnamesFromAnnotations(ocpRoute.Annotations)
+		for _, hostname := range hostnameList {
+			endpoints = append(endpoints, endpointsForHostname(hostname, targets, ttl, providerSpecific, setIdentifier)...)
+		}
+	}
+	return endpoints
+}
+
+func targetsFromOcpRouteStatus(status routeapi.RouteStatus) endpoint.Targets {
+	var targets endpoint.Targets
+
+	for _, ing := range status.Ingress {
+		if ing.Host != "" {
```

Comment:
Hi! Shouldn't this be `ing.routerCanonicalHostname` instead? it seems to me `ing.Host` will typically be the same as the route's `spec.host`, making it a not very useful target

Thanks for this PR, we're really interested in support for Openshift routes!

Reply:
Thanks for the hint. I will look into this.
I took the code from https://github.com/KohlsTechnology/external-dns/blob/timcurless-feature/source/ocproute.go and didn't wrote it myself (neither could I ;) ).
As you mentioned without an extra target annotation the generated CNAME is pointing to itself (senseless loop).
In our setup I want to use it anyway with a default target for every route, therefore I created #1495.

Reply:
@alexcern Didn't knew the RouterCanonicalHostname, but I like it! Thanks, now it makes more sense: if no target-annotation is set, the RouterCanonicalHostname will be taken. If RouterCanonicalHostname isn't set no endpoints can be created which is also fine and better than creating nonsense records.
Fixed it.

Reply:
@jgrumboe can you create some documentation for this? I think the markdown file `docs/tutorials/openshift.md` should be created to document how this works as part of this pull request.

Reply:
Done